exporter/stefexporter/exporter.go

```diff
@@ -5,26 +5,395 @@ package stefexporter // import "github.com/open-telemetry/opentelemetry-collecto
+
+	grpcWriter, opts, err := client.Connect(context.Background())
+	if err != nil {
+		s.grpcConn.Close()
+		return err
+	}
+
+	opts.Compression = s.compression
+
+	// Create record writer over gRPC stream.
+	s.stefWriter, err = oteltef.NewMetricsWriter(grpcWriter, opts)
+	if err != nil {
+		return err
+	}
+
+	// Start background timing tasks.
+	s.stopTicker = make(chan struct{})
+	go s.ticker()
+
+	s.isConnected = true
+
+	return nil
+}
+
+func (s *stefExporter) disconnect() {
+	s.connMutex.Lock()
+	defer s.connMutex.Unlock()
+
+	if !s.isConnected {
+		return
+	}
+
+	s.logger.Debug("Disconnecting...")
+
+	// Notify ticker() to stop.
+	close(s.stopTicker)
+
+	// Interrupt any pending waitForAck() call. They can't succeed anymore.
+	s.ackDataMutex.Lock()
+	s.interruptAckWaits = true
+	s.ackExpectedBy = s.ackExpectedBy[:0]
+	s.ackCond.Broadcast()
+	s.ackDataMutex.Unlock()
+
+	// If there is an existing connection close it.
+	if s.grpcConn != nil {
+		if err := s.grpcConn.Close(); err != nil {
+			s.logger.Error("failed to close grpc connection", zap.Error(err))
+		}
+	}
+
+	// Wait for all in progress wait in goroutines to get the notification that
+	// we are disconnecting and exit.
+	s.inProgressWaits.Wait()
+
+	s.isConnected = false
+}
+
 func (s *stefExporter) Shutdown(_ context.Context) error {
+	s.disconnect()
 	return nil
 }
 
-func (s *stefExporter) pushMetrics(_ context.Context, _ pmetric.Metrics) error {
+func (s *stefExporter) exportMetrics(ctx context.Context, md pmetric.Metrics) error {
+	// Keep count of how many concurrent writes are in progress.
+	s.writesInProgress.Add(1)
+	defer s.writesInProgress.Add(-1)
+
+	// Write the metrics to STEF stream.
+	expectedAckID, err := s.writeMetrics(ctx, md)
+	if err != nil {
+		return err
+	}
+
+	// Wait for acknowledgement from destination.
+	if err := s.waitForAck(expectedAckID); err != nil {
+		return err
+	}
+
 	return nil
 }
+
+func (s *stefExporter) writeMetrics(ctx context.Context, md pmetric.Metrics) (uint64, error) {
+	if err := s.ensureConnected(ctx, s.host); err != nil {
+		return 0, err
+	}
+
+	// stefWriter is not safe for concurrent writing, protect it.
+	s.stefWriterMutex.Lock()
+	defer s.stefWriterMutex.Unlock()
+
+	converter := stefpdatametrics.OtlpToTEFUnsorted{}
+	err := converter.WriteMetrics(md, s.stefWriter)
+	if err != nil {
+		// TODO: check if err is because STEF encoding failed. If so we must not
+		// try to re-encode the same data. Return consumererror.NewPermanent(err)
+		// to the caller.
+		s.disconnect()
+
+		// Return an error to retry sending these metrics again next time.
+		return 0, err
+	}
+
+	// According to STEF gRPC spec the destination ack IDs match written record number.
+	// When the data we have just written is received by destination it will send us
+	// back and ack ID that numerically matches the last written record number.
+	expectedAckID := s.stefWriter.RecordCount()
+
+	if s.writesInProgress.Load() >= s.maxWritesInProgress {
```

Comment:
Why do we send the data instead of blocking until room to send? Also does Flush mean `writesInProgress` will decrease? 

Also, you have a bug here, because you will endup with sending 1 request then flush after every request in case of high load.

Think if you have 100000 of requests that do `writesInProgress + 1` this condition will be true after each execution.

Reply:
Eliminated, no longer relevant.